File: trunk/src/org/simbrain/network/gui/actions/SetSynapsePropertiesAction.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.gui.actions;

import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.KeyEvent;

import javax.swing.AbstractAction;
import javax.swing.JComponent;
import javax.swing.KeyStroke;

import org.simbrain.network.gui.NetworkPanel;
import org.simbrain.network.gui.NetworkSelectionEvent;
import org.simbrain.network.gui.NetworkSelectionListener;

/**
 * Set synapse properties.
 */
public final class SetSynapsePropertiesAction extends AbstractAction {

    /** Network panel. */
    private final NetworkPanel networkPanel;

    /**
     * Create a new set synapse properties action with the specified network
     * panel.
     *
     * @param networkPanel networkPanel, must not be null
     */
    public SetSynapsePropertiesAction(final NetworkPanel networkPanel) {

        super("Synapse Properties...");

        if (networkPanel == null) {
            throw new IllegalArgumentException("networkPanel must not be null");
        }
        this.putValue(ACCELERATOR_KEY, KeyStroke.getKeyStroke(KeyEvent.VK_E,
                Toolkit.getDefaultToolkit().getMenuShortcutKeyMask()));
        networkPanel.getInputMap(JComponent.WHEN_IN_FOCUSED_WINDOW).put(
                KeyStroke.getKeyStroke('e'), this);

        this.networkPanel = networkPanel;
        updateAction();

        // add a selection listener to update state based on selection
        networkPanel.addSelectionListener(new NetworkSelectionListener() {

            /** @see NetworkSelectionListener */
            public void selectionChanged(final NetworkSelectionEvent event) {
                updateAction();
            }
        });
    }

    /**
     * Set action text based on number of selected neurons.
     */
    private void updateAction() {
        int numSynapses = networkPanel.getSelectedSynapses().size();

        if (numSynapses > 0) {
            String text = new String(
                    ("Edit " + numSynapses + ((numSynapses > 1) ? " Selected Synapses"
                            : " Selected Synapse")));
            putValue(NAME, text);
            setEnabled(true);
        } else {
            putValue(NAME, "Edit Selected Synapse(s)");
            setEnabled(false);
        }
    }

    /** @see AbstractAction */
    public void actionPerformed(final ActionEvent event) {

        networkPanel.showSelectedSynapseProperties();

    }
}

File: src/main/kotlin/org/simbrain/network/gui/actions/synapse/SetSynapsePropertiesAction.kt
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.gui.actions.synapse

import org.simbrain.network.core.Synapse
import org.simbrain.network.gui.NetworkPanel
import org.simbrain.network.gui.showSelectedSynapseProperties
import java.awt.Toolkit
import java.awt.event.ActionEvent
import java.awt.event.KeyEvent
import javax.swing.AbstractAction
import javax.swing.Action
import javax.swing.JComponent
import javax.swing.KeyStroke

/**
 * Set synapse properties.
 */
class SetSynapsePropertiesAction(val networkPanel: NetworkPanel) : AbstractAction("Synapse Properties...") {

    /**
     * Set action text based on number of selected neurons.
     */
    private fun updateAction() {
        val numSynapses = networkPanel.selectionManager.filterSelectedModels<Synapse>().size
        isEnabled = if (numSynapses > 0) {
            putValue(Action.NAME, "Edit $numSynapses Selected ${if (numSynapses > 1) "Synapses" else "Synapse"}")
            true
        } else {
            putValue(Action.NAME, "Edit Selected Synapse(s)")
            false
        }
    }

    override fun actionPerformed(event: ActionEvent) {
        networkPanel.showSelectedSynapseProperties()
    }

    init {
        putValue(Action.ACCELERATOR_KEY, KeyStroke.getKeyStroke(KeyEvent.VK_E, Toolkit.getDefaultToolkit().menuShortcutKeyMask))
        networkPanel.getInputMap(JComponent.WHEN_IN_FOCUSED_WINDOW).put(KeyStroke.getKeyStroke('e'), this)
        putValue(Action.SHORT_DESCRIPTION, "Set the properties of selected synapses")
        updateAction()

        // add a selection listener to update state based on selection
        networkPanel.selectionManager.events.onSelection { _, _ -> updateAction() }
    }
}